rfcs/2021-11-03-9862-ingest-apm-stats-along-traces-in-dd-agent-source.md

```diff
@@ -0,0 +1,292 @@
+are also [heavily based on ddsketch][vector-sketch], APM stats sketches can be converted to/from the Vector internal
+representation without incuring too much accuracy losses, but this would require significant work to implement those
+conversion.
+
+This opens two major very different paths for APM stats in Vector:
+
+* Either the APM stats are emitted as a log, each [`ClientGroupedStats`][client-grouped-stats-proto] would be mapped to
+  one log event.
+* Or they are emitted as metrics each value from every [`ClientGroupedStats`][client-grouped-stats-proto] would be
+  emitted as a metric with all upper level information store as tags, but as each metric event flows independently from
+  other ones this would imply adding a significant re-aggregation logic in the traces sink.
+
+Those to approach can be mixed up together if we introduce one of the following ability to bundle multiple metrics into
+a single event:
+
+* Allow multiple metric samples into a single metric event
+* Allow log events to embed metrics, it would mean adding a `Metric` type to the `Value` enum
+
+This also raises the question of having a second (assuming that the `datadog-agent` sources accepts Datadog Agent
+metrics - [RFC][dd-agent-metric-rfc]), unrelated, metric stream coming out of the `datadog-agent` source. Obviously
+event ingested representing APM stats will have to be routed along with traces, and most often they will follow a
+different path that other plain metrics/logs received from a core agent. Thus it is suggested to re-arrange the
+`datadog-agent` source, many option are available (Additional details on what exactly are "Datadog Agents" can be
+found in the trace support [RFC][trace-support-pr] and may provide relevant context for undermentionned points):
+
+* Keeping a single `datadog-agent` source:
+  * and add a settings to switch between agent kind: `agent: <TYPE>` where `<TYPE>`
+  could be  `core` (would support metrics & logs - we could optionnally add `logs` & `metrics` to only allow logs or
+  metrics, along with `core` that would allow both), `trace` and could be extended to support `process`, `security` and
+  so on
+  * or introduce the ability for a source to have multiple output for the `datadog-agent` source, e.g. <SRC_ID>.metrics, <SRC_ID>.logs, <SRC_ID>.traces, <SRC_ID>.apm_stats, etc.
+* Another solution would be to keep one Vector source per kind of Datadog Agent an then we would have the following
+  Vector sources:
+  * `datadog-core` or the current `datadog-agent` sources that would receive the data sent by the Datadog "core" Agent
+    (the agent that collects logs & metrics).
+  * `datadog-trace` that would support all data sent by the `trace-agent`
+  * And so on as the support list grows:
+    * `datadog-process` source for the Datadog `process-agent`
+    * `datadog-security` for the Datadog security Agent
+    * Etc.
```

Comment:
```suggestion
* Keeping a single `datadog_agent` source:
  * and add a settings to switch between agent kind: `agent: <TYPE>` where `<TYPE>`
  could be  `core` (would support metrics & logs - we could optionally add `logs` & `metrics` to only allow logs or
  metrics, along with `core` that would allow both), `trace` and could be extended to support `process`, `security` and
  so on
  * or introduce the ability for a source to have multiple output for the `datadog-agent` source, e.g. <SRC_ID>.metrics, <SRC_ID>.logs, <SRC_ID>.traces, <SRC_ID>.apm_stats, etc.
* Another solution would be to keep one Vector source per kind of Datadog Agent an then we would have the following
  Vector sources:
  * `datadog_core` or the current `datadog_agent` sources that would receive the data sent by the Datadog "core" Agent
    (the agent that collects logs & metrics).
  * `datadog_trace` that would support all data sent by the `trace-agent`
  * And so on as the support list grows:
    * `datadog_process` source for the Datadog `process-agent`
    * `datadog_security` for the Datadog security Agent
    * Etc.
```